internal/campaigns/volume_workspace.go

```diff
@@ -0,0 +1,199 @@
+	"github.com/sourcegraph/src-cli/internal/exec"
+)
+
+// dockerVolumeWorkspaceCreator creates dockerWorkspace instances.
+type dockerVolumeWorkspaceCreator struct {
+	client api.Client
+}
+
+var _ WorkspaceCreator = &dockerVolumeWorkspaceCreator{}
+
+func (wc *dockerVolumeWorkspaceCreator) Create(ctx context.Context, repo *graphql.Repository) (Workspace, error) {
+	w := &dockerVolumeWorkspace{}
+	return w, w.init(ctx, wc.client, repo)
+}
+
+func (*dockerVolumeWorkspaceCreator) DockerImages() []string {
+	return []string{dockerWorkspaceImage}
+}
+
+// dockerVolumeWorkspace workspaces are placed on Docker volumes (surprise!),
+// and are therefore transparent to the host filesystem. This has performance
+// advantages if bind mounts are slow, such as on Docker for Mac, but could make
+// debugging harder and is slower when it's time to actually retrieve the diff.
+type dockerVolumeWorkspace struct {
+	volume string
+}
+
+var _ Workspace = &dockerVolumeWorkspace{}
+
+func (w *dockerVolumeWorkspace) init(ctx context.Context, client api.Client, repo *graphql.Repository) error {
+	// Download the ZIP archive to a temporary file.
+	f, err := ioutil.TempFile(os.TempDir(), "src-archive-*.zip")
+	if err != nil {
+		return errors.Wrap(err, "creating temporary archive")
+	}
+	hostZip := f.Name()
+	defer os.Remove(hostZip)
+
+	if err := fetchRepositoryArchive(ctx, client, repo, hostZip); err != nil {
+		return errors.Wrap(err, "fetching repository archive")
+	}
+
+	// Create a Docker volume.
+	out, err := exec.CommandContext(ctx, "docker", "volume", "create").CombinedOutput()
+	if err != nil {
+		return errors.Wrap(err, "creating Docker volume")
+	}
+	w.volume = string(bytes.TrimSpace(out))
+
+	// Now mount that temporary file into a Docker container, and unzip it into
+	// the volume.
+	common, err := w.DockerRunOpts(ctx, "/work")
+	if err != nil {
+		return errors.Wrap(err, "generating run options")
+	}
+
+	opts := append([]string{
+		"run",
+		"--rm",
+		"--init",
+		"--workdir", "/work",
+		"--mount", "type=bind,source=" + hostZip + ",target=/tmp/zip,ro",
+	}, common...)
+	opts = append(opts, dockerWorkspaceImage, "unzip", "/tmp/zip")
+
+	out, err = exec.CommandContext(ctx, "docker", opts...).CombinedOutput()
+	if err != nil {
+		return errors.Wrapf(err, "unzip output:\n\n%s\n\n", string(out))
+	}
+
+	return nil
+}
+
+func (w *dockerVolumeWorkspace) Close(ctx context.Context) error {
+	// Cleanup here is easy: we just get rid of the Docker volume.
+	return exec.CommandContext(ctx, "docker", "volume", "rm", w.volume).Run()
+}
+
+func (w *dockerVolumeWorkspace) DockerRunOpts(ctx context.Context, target string) ([]string, error) {
+	return []string{
+		"--mount", "type=volume,source=" + w.volume + ",target=" + target,
+	}, nil
+}
+
+func (w *dockerVolumeWorkspace) WorkDir() *string { return nil }
+
+func (w *dockerVolumeWorkspace) Prepare(ctx context.Context) error {
+	script := `#!/bin/sh
+	
+set -e
+set -x
```

Comment:
Where do users see the output of `set -x`? Only in the case of error, right?

Reply:
Yep, which I think is also the only case we'll be interested in it.